Platform/OpenNtfsDxe/Compression.c

```diff
@@ -0,0 +1,685 @@
+/** @file
+  Copyright (c) 2022, Mikhail Krichanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+
+  Functional and structural descriptions follow NTFS Documentation
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+#define IS_COMPRESSED_BLOCK 0x8000
+#define BLOCK_LENGTH_BITS   0xFFF
+#define UNIT_MASK           0xF
+
+extern UINTN  mFileRecordSize;
+extern UINTN  mSectorSize;
+extern UINTN  mClusterSize;
+extern UINT64 mUnitSize;
+STATIC UINT64 mBufferSize;
+
+STATIC
+EFI_STATUS
+GetNextCluster (
+  IN OUT COMPRESSED *Clusters
+  )
+{
+  EFI_STATUS   Status;
+
+  if (Clusters->Head >= Clusters->Tail) {
+    DEBUG ((DEBUG_INFO, "NTFS: Compression block overflown\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  Status = DiskRead (
+    Clusters->FileSystem,
+    (Clusters->Elements[Clusters->Head].Lcn - Clusters->Elements[Clusters->Head].Vcn + Clusters->CurrentVcn) * mClusterSize,
+    mClusterSize,
+    Clusters->Cluster
+    );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  ++Clusters->CurrentVcn;
+
+  if (Clusters->CurrentVcn >= Clusters->Elements[Clusters->Head].Vcn) {
+    ++Clusters->Head;
+  }
+
+  Clusters->ClusterOffset = 0;
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+GetDataRunByte (
+  IN  COMPRESSED *Clusters,
+  OUT UINT8      *Result
+  )
+{
+  EFI_STATUS   Status;
+
+  ASSERT (Clusters != NULL);
+  ASSERT (Result   != NULL);
+
+  if (Clusters->ClusterOffset >= mClusterSize) {
+    Status = GetNextCluster (Clusters);
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+  }
+
+  *Result = Clusters->Cluster[Clusters->ClusterOffset++];
```

Comment:
Referencing: https://github.com/acidanthera/OpenCorePkg/pull/332/files#r851212625
The check above suggests the assumed invariant ``Clusters->ClusterOffset >= mClusterSize`` does not necessarily hold. Incrementing and access here happens outside the condition, so we can violate the invariant and access the cluster as per my assumptions. Why is this not the case?

Reply:
Sorry, forgot to comment, that was another case of "this is fixed up later"